File: problems/0005_longest-palindromic-substring/manacher/Solution.php
<?php
class Solution {
    /**
     * @param String $s
     * @return String
     */
    function longestPalindrome($s) {
        $sLen = strlen($s);
        if ($sLen <= 1)
            return $s;
        $size = 2 * $sLen + 1;
        $t = [];
        for ($i = 0; $i < $sLen; ++$i) {
            $j = 2 * $i;
            $t[$j] = 0;
            $t[$j + 1] = $s[$i];
        }
        $t[$size - 1] = 0;
        $pLens = [];
        $c = 0;
        $r = 0;
        $maxPLen = 0;
        $idx = 0;
        for ($i = 0; $i < $size; ++$i) {
            $j = 2 * $c - $i;
            $pLen = ($r > $i ?
                    min($r - $i, $pLens[$j]) : 0);
            $posLeft = $i - ($pLen + 1);
            $posRight = $i + ($pLen + 1);
            while (0 <= $posLeft && $posRight < $size &&
                    $t[$posLeft] == $t[$posRight]) {
                ++$pLen;
                --$posLeft;
                ++$posRight;
            }
            $newR = $i + ($pLens[] = $pLen);
            if ($newR > r) {
                $c = $i;
                $r = $newR;
            }
            if ($pLen > $maxPLen) {
                $maxPLen = $pLen;
                $idx = $i;
            }
        }
        $start = intdiv($idx - $maxPLen, 2);
        return substr($s, $start, $maxPLen);
    }
}

File: problems/0005_longest-palindromic-substring/manacher/Solution.kt
class Solution {
    fun longestPalindrome(s: String): String {
        if (s.length <= 1)
            return s
        val size = 2 * s.length + 1;
        val t = CharArray(size)
        for (i in 0..(s.length - 1))
            t[2 * i + 1] = s[i]
        val pLens = IntArray(size)
        var c = 0
        var r = 0
        var maxPLen = 0
        var idx = 0
        for (i in 0..(size - 1)) {
            val j = 2 * c - i;
            var pLen = if (r > i) Math.min(r - i, pLens[j])
                       else 0
            var posLeft = i - (pLen + 1)
            var posRight = i + (pLen + 1)
            while (0 <= posLeft && posRight < size &&
                    t[posLeft] == t[posRight]) {
                ++pLen
                --posLeft
                ++posRight
            }
            pLens[i] = pLen
            val newR = i + pLen
            if (newR > r) {
                c = i
                r = newR
            }
            if (pLen > maxPLen) {
                maxPLen = pLen
                idx = i
            }
        }
        val start = (idx - maxPLen) / 2
        return s.substring(start, start + maxPLen)
    }
}
